src/storage.rs

```diff
@@ -0,0 +1,151 @@
+where
+    U: Sub<U, Output = U> + Copy,
+{
+    type Output = Address<U>;
+
+    fn sub(self, other: &'b AddressOffset<U>) -> Address<U> {
+        Address(self.0 - other.0)
+    }
+}
+
+/// Page represents an unsigned integer that is a Page ID in the device memory space.
+pub struct Page<U>(pub U);
+
+/// Read a single word from the device.
+///
+/// `Word` type allows any word size to be used.
+pub trait SingleRead<Word, U> {
+    /// An enumeration of Storage errors
+    type Error;
+
+    /// Reads the word stored at the address
+    /// ```
+    /// pub fn try_read(&mut self, address: Address) -> nb::Result<u8, Self::Error>
+    ///     let address = address.0 as *const _;
+    ///     unsafe {
+    ///         Ok(core::slice::from_raw_parts::<'static, u8>(address,length))
+    ///     }
+    /// ```
+    fn try_read(&mut self, address: Address<U>) -> nb::Result<Word, Self::Error>;
+}
+
+/// Write a single word to the device.
+///
+/// `Word` type allows any word size to be used.
+pub trait SingleWrite<Word, U> {
+    /// An enumeration of Storage errors
+    type Error;
+
+    /// Writes the word to the address
+    fn try_write(&mut self, address: Address<U>, word: Word) -> nb::Result<(), Self::Error>;
+}
+
+/// Read multiple bytes from the device.
+///
+/// intended to be used for when there is a optimized method of reading multiple bytes.
+///
+/// Iterating over the slice is a valid method to ```impl``` this trait.
+pub trait MultiRead<Word, U> {
+    /// An enumeration of Storage errors
+    type Error;
+
+    /// Reads the words stored at the address to fill the buffer
+    /// ```
+    /// pub fn try_read_slice(
+    ///     &mut self,
+    ///     address: Address,  
+    ///     buf: &mut [Word]
+    /// ) -> nb::Result<Word, Self::Error>
+    ///     let address = address.0 as *const _;
+    ///     unsafe {
+    ///         buf = core::slice::from_raw_parts::<'static, Word>(address,buf.len())
+    ///     }
+    ///     
+    ///     Ok()
+    /// }
+    /// ```
+    fn try_read_slice(
+        &mut self,
+        address: Address<U>,
+        buf: &mut [Word],
+    ) -> nb::Result<(), Self::Error>;
+}
+
+/// Write multiple bytes to the device.
+///
+/// intended to be used for when there is a optimized method of reading multiple bytes.
+///
+/// Iterating over the slice is a valid method to ```impl``` this trait.
+pub trait MultiWrite<Word, U> {
+    /// An enumeration of Storage errors
+    type Error;
+
+    /// Writes the buffer to the address.
+    // Impls using spi will need a mutable buffer
+    fn try_write_slice(
+        &mut self,
+        address: Address<U>,
+        buf: &mut [Word],
+    ) -> nb::Result<(), Self::Error>;
+}
+
+/// A common interface to erase pages or memory locations.
+///
+/// For Flash storage, the write functions can't set a bit to 1.
+///
+/// For non flash devices, this trait is not required, but it can be used to erase data as recommended by the device (EG set all fields to 0).
```

Comment:
If the behavior of these functions is device dependent, this trait is not really useful for generic drivers, because the driver has no way of knowing what the ErasePage trait will do. By driver I mean a file system implementation for example. That's why I proposed distinct traits for different kinds of storage with well defined behavior.